src/cookbook/games/firestore-multiplayer.md

```diff
@@ -0,0 +1,469 @@
+    that was generated by `flutterfire configure` in the previous step.
+
+    ```dart
+    import 'package:cloud_firestore/cloud_firestore.dart';
+    import 'package:firebase_core/firebase_core.dart';
+    
+    import 'firebase_options.dart';
+    ```
+
+2.  Add the following code just above the call to `runApp()`
+    in `lib/main.dart`:
+
+    ```dart
+     WidgetsFlutterBinding.ensureInitialized();
+    
+      await Firebase.initializeApp(
+        options: DefaultFirebaseOptions.currentPlatform,
+      );
+    ```
+
+    This ensures that Firebase is initialized on game startup.
+
+3.  Add the Firestore instance to the app.
+    That way, any widget can access this instance.
+    Widgets can also react to the instance missing if needed.
+
+    Replace the usual `runApp(MyApp())` with the following:
+
+    ```dart
+    runApp(
+      Provider.value(
+        value: FirebaseFirestore.instance,
+        child: MyApp(),
+      ),
+    );
+    ```
+
+    Put the provider above `MyApp`, not inside it.
+    This enables you to test the app without Firebase.
+
+
+## 4. Create a Firestore controller class
+
+Though you can talk to Firestore directly,
+you should write a dedicated controller class
+to make the code more readable and maintainable.
+
+How you implement the controller depends on your game
+and on the exact design of your multiplayer experience.
+For the case of the `card` template,
+you could synchronize the contents of the two circular playing areas.
+It's not enough for a full multiplayer experience, but it's a good start.
+
+![Screenshot of the card game, with arrows pointing to playing areas](/assets/images/docs/cookbook/multiplayer-areas.jpg){:.site-mobile-screenshot .site-illustration}
+
+To create a controller, copy, then paste the following code into a new file:
+`lib/multiplayer/firestore_controller.dart`.
+
+```dart
+import 'dart:async';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:flutter/foundation.dart';
+import 'package:logging/logging.dart';
+
+import '../game_internals/board_state.dart';
+import '../game_internals/playing_area.dart';
+import '../game_internals/playing_card.dart';
+
+class FirestoreController {
+  static final _log = Logger('FirestoreController');
+
+  final FirebaseFirestore instance;
+
+  final BoardState boardState;
+
+  /// For now, there is only one match. But in order to be ready
+  /// for match-making, put it in a Firestore collection called matches.
+  late final _matchRef = instance.collection('matches').doc('match_1');
+
+  late final _areaOneRef = _matchRef
+      .collection('areas')
+      .doc('area_one')
+      .withConverter<List<PlayingCard>>(
+          fromFirestore: _cardsFromFirestore, toFirestore: _cardsToFirestore);
+
+  late final _areaTwoRef = _matchRef
+      .collection('areas')
+      .doc('area_two')
+      .withConverter<List<PlayingCard>>(
+          fromFirestore: _cardsFromFirestore, toFirestore: _cardsToFirestore);
```

Comment:
Not 100% necessary, just a question. Would it be useful to add the explicit types here? I generally do it for fields as a best practice, but it also might help readers understand what's being returned since this is on a web page rather than IDE.

Reply:
I'm going to take a note and revisit this later. I see your point but I'm torn. The Dart style is not to add extraneous types to private members. On the other hand, this particular piece of code _is_ dense — I can totally see people wanting to have the types explicit there.

Reply:
I'd lean against over specifying the types here. Looking to see what the community is doing lead to this S/O post:

https://stackoverflow.com/questions/68079030/how-to-use-firestore-withconverter-in-flutter

I think this code is good as is. =)

Reply:
Works for me :)